File: v16_bridge/java/OnlineStore2.java
package v16_bridge.java;

interface Payment {
    void processPayment();
}

class CreditCardPayment implements Payment {
    public void processPayment() {
        System.out.println("Processing credit card payment");
    }
}

class WeChatPayment implements Payment {
    public void processPayment() {
        System.out.println("Processing wechat payment");
    }
}

abstract class Product {
    protected Payment payment;
    public Product(Payment payment) {
        this.payment = payment;
    }
    public abstract void purchase();
}

class Book extends Product {
    public Book(Payment payment) {
        super(payment);
    }
    public void purchase() {
        this.payment.processPayment();
        System.out.println("Purchased book");
    }
}

class Electronics extends Product {
    public Electronics(Payment payment) {
        super(payment);
    }
    public void purchase() {
        this.payment.processPayment();
        System.out.println("Purchased electronics");
    }
}


public class OnlineStore2 {
    public static void main(String[] args) {

        Payment creditCardPayment = new CreditCardPayment();
        Payment wechatPayment = new WeChatPayment();
        Book book = new Book(creditCardPayment);
        book.purchase();
        Electronics electronics = new Electronics(creditCardPayment);
        electronics.purchase();
        Book book2 = new Book(wechatPayment);
        book2.purchase();
    }
}

File: v16_bridge/python/online_store2.py
from abc import ABC, abstractmethod

class Payment(ABC):
    @abstractmethod
    def processPayment(self):
        pass

class CreditCardPayment(Payment):
    def processPayment(self):
        print("Processing credit card payment")

class WeChatPayment(Payment):
    def processPayment(self):
        print("Processing wechat payment")

class Product(ABC):
    def __init__(self, payment: Payment):
        self.payment = payment

    @abstractmethod
    def purchase(self):
        pass

class Book(Product):
    def __init__(self, payment: Payment):
        super().__init__(payment)

    def purchase(self):
        self.payment.processPayment()
        print("Purchased book")

class Electronics(Product):
    def __init__(self, payment: Payment):
        super().__init__(payment)

    def purchase(self):
        self.payment.processPayment()
        print("Purchased electronics")

if __name__ == "__main__":
    creditCardPayment = CreditCardPayment()
    wechatPayment = WeChatPayment()
    book = Book(creditCardPayment)
    book.purchase()
    electronics = Electronics(creditCardPayment)
    electronics.purchase()
    book2 = Book(wechatPayment)
    book2.purchase()


